### ISSUE TITLE
Server pod was stuck in terminating state after deleting it in a k3d cluster

### ISSUE THREAD

**Issue by leelavg** [kadalu/kadalu#448]
- Using the similar setup as mentioned in [other](https://github.com/kadalu/kadalu/issues/441#issue-807207824) issue and performing below operation resulted in the Server Pod stuck in terminating state
- `cp` of /usr to /mnt/gluster and it's arequal checksum matched
- `cp` of /sys to /mnt/gluster + kill a server pod is stuck (in terminating state forever: >1hr), any command related to mount (df/ls on mnt) is also stuck,  below warning line is spamming logs
```
[2021-02-16 11:25:15.830574 +0000] W [socket.c:764:__socket_rwv] 0-tcp.replica3-pool-server: readv on 10.42.1.1:50836 failed (No data available)
[2021-02-16 11:25:19.172335 +0000] I [addr.c:52:compare_addr_and_update] 0-/bricks/replica3-pool/data/brick: allowed = "*", received addr = "10.42.3.7"
[2021-02-16 11:25:19.172471 +0000] I [MSGID: 115029] [server-handshake.c:564:server_setvolume] 0-replica3-pool-server: accepted client from CTX_ID:f49f0b24-9e40-4fda-af4f-ec827f38663e-GRAPH_ID:0-PID:4562-HOST:k
adalu-csi-provisioner-0-PC_NAME:replica3-pool-client-0-RECON_NO:-0 (version: 2021.02.13) with subvol /bricks/replica3-pool/data/brick
[2021-02-16 11:25:19.227333 +0000] W [socket.c:764:__socket_rwv] 0-tcp.replica3-pool-server: readv on 10.42.3.7:49148 failed (No data available)
[2021-02-16 11:25:19.227435 +0000] I [MSGID: 115036] [server.c:500:server_rpc_notify] 0-replica3-pool-server: disconnecting connection [{client-uid=CTX_ID:f49f0b24-9e40-4fda-af4f-ec827f38663e-GRAPH_ID:0-PID:456
2-HOST:kadalu-csi-provisioner-0-PC_NAME:replica3-pool-client-0-RECON_NO:-0}]
[2021-02-16 11:25:19.227523 +0000] W [MSGID: 115020] [server.c:505:server_rpc_notify] 0-replica3-pool-server: failed to get [{type=auth-path}]
[2021-02-16 11:25:19.227919 +0000] I [MSGID: 101055] [client_t.c:396:gf_client_unref] 0-replica3-pool-server: Shutting down connection CTX_ID:f49f0b24-9e40-4fda-af4f-ec827f38663e-GRAPH_ID:0-PID:4562-HOST:kadalu
-csi-provisioner-0-PC_NAME:replica3-pool-client-0-RECON_NO:-0
[2021-02-16 11:25:20.103409 +0000] W [socket.c:764:__socket_rwv] 0-tcp.replica3-pool-server: readv on 10.42.1.1:50884 failed (No data available)
```
- Data stored in back-end bricks
```
(host) # for i in {c,d,e}; do du -cbhs /mnt/sd$i; done;
1.6M    /mnt/sdc
1.6M    total
1.6M    /mnt/sdd
1.6M    total
1.6M    /mnt/sde
1.6M    total
```
Ideally this amount of data should have be stored
```
(test-pod) # du -cbhs /sys
192M    /sys
192M    total
```
So, it's safe to assume only 1.6M was written before killing kadalu server pod.
- Versions deployed as below:
```
# k3d version
k3d version v4.2.0
k3s version v1.20.2-k3s1 (default)
# k apply -f https://github.com/kadalu/kadalu/releases/download/0.8.0alpha2/kadalu-operator.yaml
# k kadalu version
kubectl-kadalu 0.8.0alpha2
```

**Comment by leelavg** [comment]
Devel branch seemed to have fixed Terminating status issue

- Backend devices, no `lsblk` available in node so just mounted
```
# for i in {0,1,2}; do name=k3d-test-agent-$i; echo $name; docker exec -it $name sh -c 'df -h /mnt/sd*'; done;
k3d-test-agent-0
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdc         10G  104M  9.9G   2% /mnt/sdc
k3d-test-agent-1
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdd         10G  104M  9.9G   2% /mnt/sdd
k3d-test-agent-2
Filesystem      Size  Used Avail Use% Mounted on
/dev/sde         10G  104M  9.9G   2% /mnt/sde

# kubectl apply -f https://raw.githubusercontent.com/kadalu/kadalu/devel/manifests/kadalu-operator.yaml
# bat storage-config-device.yaml --plain; k apply -f storage-config-device.yaml 
---
apiVersion: kadalu-operator.storage/v1alpha1
kind: KadaluStorage
metadata:
  name: replica3-pool
spec:
  type: Replica3
  storage:
    - node: k3d-test-agent-0
      device: /dev/sdc
    - node: k3d-test-agent-1
      device: /dev/sdd
    - node: k3d-test-agent-2
      device: /dev/sde
kadalustorage.kadalu-operator.storage/replica3-pool created

# k exec -it test-5cfff64b66-xnbx4 -- sh -c 'cd /mnt/alpha && for i in `seq 1 9`; do cat /dev/urandom | tr -dc [:space:][:print:] | head -c 100m > $i-100mFile; echo created $i-100mFile; sleep 5 done'
created 1-100mFile
created 2-100mFile
created 3-100mFile
created 4-100mFile
created 5-100mFile
created 6-100mFile
created 7-100mFile
created 8-100mFile
created 9-100mFile
```
- pool-0-0 was deleted while above IO was happening and didn't stuck in Terminating state
```
# k get pods
NAME                          READY   STATUS    RESTARTS   AGE
operator-74c9fb55c4-xmc6k     1/1     Running   0          41m
kadalu-csi-nodeplugin-ffdtc   3/3     Running   0          41m
kadalu-csi-nodeplugin-lswhc   3/3     Running   0          41m
kadalu-csi-nodeplugin-kn72x   3/3     Running   0          41m
kadalu-csi-nodeplugin-67n8k   3/3     Running   0          41m
kadalu-csi-provisioner-0      5/5     Running   0          41m
server-replica3-pool-1-0      1/1     Running   0          40m
server-replica3-pool-2-0      1/1     Running   0          40m
test-5cfff64b66-xnbx4         1/1     Running   0          33m
server-replica3-pool-0-0      1/1     Running   0          25m
```
- Correct reflection of data used
```
# for i in {0,1,2}; do name=k3d-test-agent-$i; echo $name; docker exec -it $name sh -c 'df -h | grep pool'; done;
k3d-test-agent-0
kadalu:replica3-pool                2.1G  901M  1.2G  44% /var/lib/kubelet/pods/769c2329-5705-4ad9-9b0f-173bf79c9817/volumes/kubernetes.io~csi/pvc-d3e93914-bd66-46e0-8f09-096ec5bcee41/mount
k3d-test-agent-1
kadalu:replica3-pool                2.1G  901M  1.2G  44% /var/lib/kubelet/pods/769c2329-5705-4ad9-9b0f-173bf79c9817/volumes/kubernetes.io~csi/pvc-d3e93914-bd66-46e0-8f09-096ec5bcee41/mount
k3d-test-agent-2
kadalu:replica3-pool                2.1G  901M  1.2G  44% /var/lib/kubelet/pods/769c2329-5705-4ad9-9b0f-173bf79c9817/volumes/kubernetes.io~csi/pvc-d3e93914-bd66-46e0-8f09-096ec5bcee41/mount

# for i in {0,1,2}; do name=k3d-test-agent-$i; echo $name; docker exec -it $name sh -c 'df -h /mnt/sd*'; done;
k3d-test-agent-0
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdc         10G 1004M  9.1G  10% /mnt/sdc
k3d-test-agent-1
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdd         10G 1004M  9.1G  10% /mnt/sdd
k3d-test-agent-2
Filesystem      Size  Used Avail Use% Mounted on
/dev/sde         10G 1004M  9.1G  10% /mnt/sde

# k get pv,pvc -o wide
NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM          STORAGECLASS      REASON   AGE   VOLUMEMODE
persistentvolume/pvc-d3e93914-bd66-46e0-8f09-096ec5bcee41   2Gi        RWX            Delete           Bound    kadalu/pvc2g   kadalu.replica3            40m   Filesystem

NAME                          STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE   VOLUMEMODE
persistentvolumeclaim/pvc2g   Bound    pvc-d3e93914-bd66-46e0-8f09-096ec5bcee41   2Gi        RWX            kadalu.replica3   40m   Filesystem
```
- Devel branch seemed to have fixed #300 as well

## New issues:
- Utilisation status isn't reflected correctly
```
# k kadalu storage-list --status
Failed to get size details of the storage "replica3-pool"
error 1 Error from server (BadRequest): container glusterfsd is not valid for pod server-replica3-pool-0-0

Name             Type        Utilization            Pvs Count      Min PV Size      Avg PV Size      Max PV Size
replica3-pool    Replica3    0/0 (0%)                       0                0                0                0 
```
- Failure in `healinfo` command is observed
```
# k kadalu healinfo
Error while running the following command
$ kubectl exec -nkadalu kadalu-csi-provisioner-0 -c kadalu-provisioner -- /kadalu/heal-info.sh

error: Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "6eb517b7198346bfe8220483fc9942a0f4cbdadd26a8a917c02b786341464486": OCI runtime exec failed: exec failed: container_linux.go:370: starting container process caused: exec: "/kadalu/heal-info.sh": permission denied: unknown

# k kadalu version
kubectl-kadalu 0.8.0alpha3
```

**Comment by amarts** [comment]
> error: Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "6eb517b7198346bfe8220483fc9942a0f4cbdadd26a8a917c02b786341464486": OCI runtime exec failed: exec failed: container_linux.go:370: starting container process caused: exec: "/kadalu/heal-info.sh": permission denied: unknown

I remember seeing it before I added `chmod +x /kadalu/heal-info.sh` in the Dockerfile. Will make a successful alpha release and see if this is fixed in that particular version.

**Comment by leelavg** [comment]
```
# k kadalu healinfo
Giving heal information of volume replica3-pool
Brick server-replica3-pool-0-0.replica3-pool:/bricks/replica3-pool/data/brick
Status: Connected
Total Number of entries: 0
Number of entries in heal pending: 0
Number of entries in split-brain: 0
Number of entries possibly healing: 0

Brick server-replica3-pool-1-0.replica3-pool:/bricks/replica3-pool/data/brick
Status: Connected
Total Number of entries: 0
Number of entries in heal pending: 0
Number of entries in split-brain: 0
Number of entries possibly healing: 0

Brick server-replica3-pool-2-0.replica3-pool:/bricks/replica3-pool/data/brick
Status: Connected
Total Number of entries: 0
Number of entries in heal pending: 0
Number of entries in split-brain: 0
Number of entries possibly healing: 0
```
As above issue is fixed in latest version (0.8) closing the issue.
